modules/mempressure.c

```diff
@@ -0,0 +1,754 @@
+{
+    .mnl_used   = 0,
+};
+
+/** Cached memory use level */
+static memnotify_level_t mempressure_level = MEMNOTIFY_LEVEL_UNKNOWN;
+
+/** Check current memory status against triggering levels
+ */
+static memnotify_level_t
+mempressure_status_evaluate_level(void)
+{
+    memnotify_level_t res = MEMNOTIFY_LEVEL_UNKNOWN;
+
+    if( mempressure_limit_is_valid(&mempressure_status) ) {
+        res = MEMNOTIFY_LEVEL_NORMAL;
+
+        for( memnotify_level_t lev = MEMNOTIFY_LEVEL_NORMAL + 1;
+             lev < G_N_ELEMENTS(mempressure_limit); ++lev ) {
+            if( mempressure_limit_exceeded(mempressure_limit+lev, &mempressure_status) )
+                res = lev;
+        }
+    }
+
+    return res;
+}
+
+/** Re-evaluate memory use level and broadcast changes via datapipe
+ */
+static bool
+mempressure_status_update_level(void)
+{
+    memnotify_level_t prev = mempressure_level;
+    mempressure_level = mempressure_status_evaluate_level();
+
+    if( mempressure_level == prev )
+        goto EXIT;
+
+    mce_log(LL_WARN, "mempressure_level: %s -> %s",
+            memnotify_level_repr(prev),
+            memnotify_level_repr(mempressure_level));
+
+    datapipe_exec_full(&memnotify_level_pipe
+                       , GINT_TO_POINTER(mempressure_level));
+
+EXIT:
+
+    return mempressure_level != MEMNOTIFY_LEVEL_UNKNOWN;
+}
+
+/** Log current memory level configuration for debugging purposes
+ */
+static void
+mempressure_status_show_triggers(void)
+{
+    if( mce_log_p(LL_DEBUG) ) {
+        for( size_t i = 0; i < G_N_ELEMENTS(mempressure_limit); ++i ) {
+            char tmp[256];
+            mempressure_limit_repr(mempressure_limit+i, tmp, sizeof tmp);
+            mce_log(LL_DEBUG, "%s: %s", memnotify_level_repr(i), tmp);
+        }
+    }
+}
+
+/* ========================================================================= *
+ * MEMPRESSURE_CGROUP
+ * ========================================================================= */
+
+/** File descriptor for CGROUP_DATA_PATH */
+static int   mempressure_cgroup_data_fd  = -1;
+
+/** File descriptor for CGROUP_CTRL_PATH */
+static int   mempressure_cgroup_ctrl_fd  = -1;
+
+/** Eventfd for receiving notifications about threshold crossings */
+static int   mempressure_cgroup_event_fd = -1;
+
+/** I/O watch for mempressure_cgroup_event_fd */
+static guint mempressure_cgroup_event_id = 0;
+
+/** Probe if the required cgroup sysfs files are present
+ */
+static bool
+mempressure_cgroup_is_available(void)
+{
+    return (access(CGROUP_DATA_PATH, R_OK) == 0 &&
+            access(CGROUP_CTRL_PATH, W_OK) == 0);
+}
+
+/** Input watch callback for cgroup memory threshold crossings
+ */
+static gboolean
+mempressure_cgroup_event_cb(GIOChannel *chn, GIOCondition cnd, gpointer aptr)
+{
+    (void) chn;
+    (void) aptr;
+
+    gboolean ret = G_SOURCE_REMOVE;
+
+    if( !mempressure_cgroup_event_id )
+        goto EXIT;
+
+    if( mempressure_cgroup_event_fd == -1 )
```

Comment:
Would it make sense to compare these to one checking (mempressure_cgroup_event_id <= 0)?